src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs

```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessageMethodTypeParameterMoreMeaningfulName,
+            DiagnosticCategory.Naming,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: false,
+            description: s_localizableDescription,
+            helpLinkUri: "https://docs.microsoft.com/visualstudio/code-quality/ca1704",
+            customTags: FxCopWellKnownDiagnosticTags.PortedFxCopRule);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            FileParseRule,
+            AssemblyRule,
+            NamespaceRule,
+            TypeRule,
+            VariableRule,
+            MemberRule,
+            MemberParameterRule,
+            DelegateParameterRule,
+            TypeTypeParameterRule,
+            MethodTypeParameterRule,
+            AssemblyMoreMeaningfulNameRule,
+            NamespaceMoreMeaningfulNameRule,
+            TypeMoreMeaningfulNameRule,
+            MemberMoreMeaningfulNameRule,
+            MemberParameterMoreMeaningfulNameRule,
+            DelegateParameterMoreMeaningfulNameRule,
+            TypeTypeParameterMoreMeaningfulNameRule,
+            MethodTypeParameterMoreMeaningfulNameRule);
+
         public override void Initialize(AnalysisContext analysisContext)
-#pragma warning restore RS1025 // Configure generated code analysis
         {
             analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationStartContext)
+        {
+            var dictionaries = ReadDictionaries();
+            var projectDictionary = CodeAnalysisDictionary.CreateFromDictionaries(dictionaries.Concat(_mainDictionary));
+
+            compilationStartContext.RegisterOperationAction(AnalyzeVariable, OperationKind.VariableDeclarator);
+            compilationStartContext.RegisterCompilationEndAction(AnalyzeAssembly);
+            compilationStartContext.RegisterSymbolAction(
+                AnalyzeSymbol,
+                SymbolKind.Namespace,
+                SymbolKind.NamedType,
+                SymbolKind.Method,
+                SymbolKind.Property,
+                SymbolKind.Event,
+                SymbolKind.Field,
+                SymbolKind.Parameter);
+
+            IEnumerable<CodeAnalysisDictionary> ReadDictionaries()
+            {
+                var fileProvider = AdditionalFileProvider.FromOptions(compilationStartContext.Options);
+                return fileProvider.GetMatchingFiles(@"(?:dictionary|custom).*?\.(?:xml|dic)$")
+                    .Select(CreateDictionaryFromAdditionalText)
+                    .Where(x => x != null)
+                    .ToList();
```

Comment:
```suggestion
                    .Where(x => x != null);
```

Reply:
Without `.ToList`,  this will be enumerated twice and cause my FileParseError test to fail (2 diagnostics)

Reply:
I don't understand why, as far as I can see it's called only once from the compilation start so it should be ok.

Reply:
I set `dictionaries` to the value of `ReadDictionaries()`, which is the enumerable. Then I pass `dictionaries.Concat(s_mainDictionary)` into `CodeAnalysisDictionary.CreateFromDictionaries`, which does the following:

```
var recognizedWords = dictionaries.Select(x => x.RecognizedWords).Aggregate((x, y) => x.Union(y));
var unrecognizedWords = dictionaries.Select(x => x.UnrecognizedWords).Aggregate((x, y) => x.Union(y));
```

As you can see, the value of `dictionaries` is enumerated twice.